src/libsyntax/codemap.rs

```diff
@@ -579,38 +579,75 @@ impl CodeMap {
 
     /// Returns a new span representing just the end-point of this span
     pub fn end_point(&self, sp: Span) -> Span {
-        let hi = sp.hi().0.checked_sub(1).unwrap_or(sp.hi().0);
-        let hi = self.get_start_of_char_bytepos(BytePos(hi));
-        let lo = cmp::max(hi.0, sp.lo().0);
-        sp.with_lo(BytePos(lo))
+        let pos = sp.hi().0;
+
+        let width = self.find_width_of_character_at_span(sp, false);
+        let corrected_end_position = pos.checked_sub(width).unwrap_or(pos);
+
+        let end_point = BytePos(cmp::max(corrected_end_position, sp.lo().0));
+        sp.with_lo(end_point)
     }
 
     /// Returns a new span representing the next character after the end-point of this span
     pub fn next_point(&self, sp: Span) -> Span {
-        let hi = sp.lo().0.checked_add(1).unwrap_or(sp.lo().0);
-        let hi = self.get_start_of_char_bytepos(BytePos(hi));
-        let lo = cmp::max(sp.hi().0, hi.0);
-        Span::new(BytePos(lo), BytePos(lo), sp.ctxt())
+        let pos = sp.lo().0;
+
+        let width = self.find_width_of_character_at_span(sp, true);
+        let corrected_next_position = pos.checked_add(width).unwrap_or(pos);
+
+        let next_point = BytePos(cmp::max(sp.hi().0, corrected_next_position));
```

Comment:
You also need to account for the (low) chance that `next_point` might _also_ be a wide character, as `next_point` could potentially be used anywhere, including at the start of an ident (in practice this might never happen, and even if it does, the presentation would be ok, but would break havoc on tools depending on offsets).

Reply:
Does the second parameter of `find_width_of_character_at_span` that searches forward for the character boundary rather than backwards not handle this or am I misunderstanding?

Reply:
My understanding of this code is that in the code `println!("☃☃")` if you have a span pointing at `println!(` and use `next_point`, the new span will point at the first `"`. If you call `next_point` on _that_ span, it would point to `0xE2`, when it should actually be pointing at `0xE2 0x98 0x83`. This happens because of the line below, where you create the span with the same start and end. Your code correctly handles the case where your span points to the inside of the text `0xE2 0x98 0x83 0xE2 0x98 0x83` and you call `next_point`, yielding the second `"`. Does that make sense?

Reply:
In the original version of the `next_point` function it returned a span with the same start and end, wouldn't it have had the same issue?

Reply:
Yes :)

Reply:
I'm not sure I understand, what change is required?

Reply:
We can do that on a follow up PR. I'll approve once ci is happy.

Reply:
I've already got it added. Only had one little thing remaining yesterday but it was getting late.

Reply:
No problem! Thank you for all the work you put into this! I know that the user facing change is  not that big given the effort, but you're fixing quite a few potential pernicious ICEs :)

Reply:
Pushed up the fix for this.